compiler/testData/diagnostics/tests/enum/incompatibleEnums_withExtendedTypeCompatibilityChecker.fir.kt

```diff
@@ -0,0 +1,145 @@
+    <!INCOMPATIBLE_ENUM_COMPARISON_ERROR!>E1.B == e2<!>
+
+    <!INCOMPATIBLE_ENUM_COMPARISON_ERROR!>E1.A == E2.B<!>
+
+    e1 == E1.A
+    E1.A == e1
+    e2 == E2.B
+    E2.B == e2
+}
+
+fun foo2(e1: E1, e2: E2) {
+    when (e1) {
+        E1.A -> {}
+            <!INCOMPATIBLE_ENUM_COMPARISON_ERROR!>E2.A<!> -> {}
+            <!INCOMPATIBLE_ENUM_COMPARISON_ERROR!>E2.B<!> -> {}
+        e1 -> {}
+            <!INCOMPATIBLE_ENUM_COMPARISON_ERROR!>e2<!> -> {}
+        else -> {}
+    }
+}
+
+fun foo3(e1: Enum<E1>, e2: Enum<E2>, e: Enum<*>) {
+    e1 == e
+    <!INCOMPATIBLE_ENUM_COMPARISON_ERROR!>e1 == e2<!>
+
+    e1 == E1.A
+    <!INCOMPATIBLE_ENUM_COMPARISON_ERROR!>e1 == E2.A<!>
+
+    when (e1) {
+        e1 -> {}
+        <!INCOMPATIBLE_ENUM_COMPARISON_ERROR!>e2<!> -> {}
+        e -> {}
+        E1.A -> {}
+            <!INCOMPATIBLE_ENUM_COMPARISON_ERROR!>E2.A<!> -> {}
+        else -> {}
+    }
+
+    when (e) {
+        e -> {}
+        e2 -> {}
+        E1.A -> {}
+        E2.A -> {}
+        else -> {}
+    }
+}
+
+interface MyInterface
+open class MyOpenClass
+
+fun foo4(e1: E1, i: MyInterface, c: MyOpenClass) {
+    <!EQUALITY_NOT_APPLICABLE!>e1 == i<!>
+    <!EQUALITY_NOT_APPLICABLE!>i == e1<!>
+
+    <!EQUALITY_NOT_APPLICABLE!>e1 == c<!>
+    <!EQUALITY_NOT_APPLICABLE!>c == e1<!>
+
+    when (e1) {
+            <!INCOMPATIBLE_TYPES!>i<!> -> {}
+            <!INCOMPATIBLE_TYPES!>c<!> -> {}
+        else -> {}
+    }
+}
+
+enum class E3 : MyInterface { X, Y }
+
+fun foo5(i: MyInterface, a: Any) {
+    E3.X == E3.Y
+    E3.X == i
+    E3.X == a
+}
+
+fun foo6(e1: E1?, e2: E2) {
+    E1.A == null
+    null == E1.A
+    e1 == null
+    null == e1
+
+    <!INCOMPATIBLE_ENUM_COMPARISON_ERROR!>e1 == E2.A<!>
+    <!INCOMPATIBLE_ENUM_COMPARISON_ERROR!>E2.A == e1<!>
+    <!INCOMPATIBLE_ENUM_COMPARISON_ERROR!>e1 == e2<!>
+    <!INCOMPATIBLE_ENUM_COMPARISON_ERROR!>e2 == e1<!>
+
+    e2 == null
+    null == e2
+    E1.A == null
+    null == E1.A
+}
+
+fun foo7(e1: E1?, e2: E2?) {
+    e1 == e2 // There should be an IDE-inspection for such cases
+}
+
+fun <T> foo8(e1: E1?, e2: E2, t: T) {
+    e1 == t
+    t == e1
+
+    e2 == t
+    t == e2
+
+    E1.A == t
+    t == E1.A
+}
+
+fun <T, K> foo9(e1: E1?, e2: E2, t: T, k: K) where T : MyInterface, T : MyOpenClass, K : MyInterface {
+    <!EQUALITY_NOT_APPLICABLE!>e1 == t<!>
+    <!EQUALITY_NOT_APPLICABLE!>t == e1<!>
+
+    <!EQUALITY_NOT_APPLICABLE!>e2 == t<!>
+    <!EQUALITY_NOT_APPLICABLE!>t == e2<!>
+
+    <!EQUALITY_NOT_APPLICABLE!>E1.A == t<!>
+    <!EQUALITY_NOT_APPLICABLE!>t == E1.A<!>
+
+    <!EQUALITY_NOT_APPLICABLE!>E3.X == t<!>
+
+    E3.X == k
+    k == E3.X
+}
+
+interface Inv<T>
+
+enum class E4 : Inv<Int> { A }
+
+fun foo10(e4: E4, invString: Inv<String>) {
+    <!EQUALITY_NOT_APPLICABLE!>e4 == invString<!>
+    <!EQUALITY_NOT_APPLICABLE!>invString == e4<!>
+
+    <!EQUALITY_NOT_APPLICABLE!>E4.A == invString<!>
+    <!EQUALITY_NOT_APPLICABLE!>invString == E4.A<!>
```

Comment:
Just to double-check, these look like the cases where FE 1.0 missed?

Reply:
Yeah, it's what I think the check should have covered ideally.